flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/RecordWriterOutputWrapper.java

```diff
@@ -0,0 +1,162 @@
+
+	private RecordWriterOutputWrapper(RecordWriterOutput<?>[] recordWriterOutputs) {
+		this.recordWriterOutputs = Preconditions.checkNotNull(recordWriterOutputs);
+	}
+
+	public void broadcastCheckpointBarrier(
+			long id,
+			long timestamp,
+			CheckpointOptions checkpointOptions) throws IOException {
+		CheckpointBarrier barrier = new CheckpointBarrier(id, timestamp, checkpointOptions);
+		for (RecordWriterOutput<?> output : recordWriterOutputs) {
+			output.broadcastEvent(barrier);
+		}
+	}
+
+	public void broadcastCheckpointCancelMarker(long id) throws IOException {
+		CancelCheckpointMarker barrier = new CancelCheckpointMarker(id);
+		for (RecordWriterOutput<?> output : recordWriterOutputs) {
+			output.broadcastEvent(barrier);
+		}
+	}
+
+	/**
+	 * This method should be called before finishing the record emission, to make sure any data
+	 * that is still buffered will be sent. It also ensures that all data sending related
+	 * exceptions are recognized.
+	 *
+	 * @throws IOException Thrown, if the buffered data cannot be pushed into the output streams.
+	 */
+	public void flushOutputs() throws IOException {
+		for (RecordWriterOutput<?> output : recordWriterOutputs) {
+			output.flush();
+		}
+	}
+
+	/**
+	 * This method releases all resources of the record writer output. It stops the output
+	 * flushing thread (if there is one) and releases all buffers currently held by the
+	 * output serializers.
+	 *
+	 * <p>This method should never fail.
+	 */
+	public void releaseOutputs() {
+		for (RecordWriterOutput<?> output : recordWriterOutputs) {
+			output.close();
+		}
+	}
+
+	public RecordWriterOutput<?>[] getRecordWriterOutputs() {
+		return recordWriterOutputs;
+	}
+
+	private static RecordWriter<SerializationDelegate<StreamRecord<?>>> createRecordWriter(
+			StreamEdge edge,
+			int outputIndex,
+			Environment environment,
+			long bufferTimeout) {
+		StreamPartitioner<?> outputPartitioner = edge.getPartitioner();
+		String taskName = environment.getTaskInfo().getTaskName();
+
+		LOG.debug("Using partitioner {} for output {} of task {}", outputPartitioner, outputIndex, taskName);
+
+		ResultPartitionWriter bufferWriter = environment.getWriter(outputIndex);
+
+		// we initialize the partitioner here with the number of key groups (aka max. parallelism)
+		if (outputPartitioner instanceof ConfigurableStreamPartitioner) {
+			int numKeyGroups = bufferWriter.getNumTargetKeyGroups();
+			if (0 < numKeyGroups) {
+				((ConfigurableStreamPartitioner) outputPartitioner).configure(numKeyGroups);
+			}
+		}
+
+		RecordWriter<SerializationDelegate<StreamRecord<?>>> output = new RecordWriterBuilder()
+			.setChannelSelector(outputPartitioner)
+			.setTimeout(bufferTimeout)
+			.setTaskName(taskName)
+			.build(bufferWriter);
+		output.setMetricGroup(environment.getMetricGroup().getIOMetricGroup());
+		return output;
+	}
+
+	public static RecordWriterOutputWrapper build(StreamConfig config, Environment environment) {
```

Comment:
nit: Maybe this just should be another constructor?

Reply:
Actually I tried to provide a factory role for constructing `RecordWriterOutputWrapper`.

I guess it is not suggested involving in complex processes in the constructor, and the better way is to build them in the factory and pass the final required arguments directly into the constructor. I am not sure whether it is a common rule.

Reply:
I don't think we have any hard rule in Flink about that. Probably in most places we are not using static methods if we can use just constructor. For me, builders/factories are a necessary overhead to overcome Java's language limitation of handling multiple parameters with default values, which is not the case here, so I would use just plain constructor. On it's own factories/builders for me do not have any added value. if java's syntax supported default parameters [(like python)](https://www.geeksforgeeks.org/default-arguments-in-python/), I wouldn't be using factories/builders   

However if you have strong feelings about this, we can try it out and maybe overtime you will be able to change our practice, so feel free to ignore my preference here :)

Reply:
Thanks for this explanation. I have no personal preference for this issue. Just because I was suggested that in previous PRs. Then I can refactor this as a constructor. :)